### PR TITLE
Ofmcc 4329 show irreg expense

### PR DESCRIPTION
**Pull request by jenbeckett**
Notes: Still have some kinks to work out with Irreg Expense on the Applications Page - if MOD FA is not in Active status - then the Irreg Expenses don't show up 

Irregular Expense PDF notification doesn't generate yet on the Funding Tab

### CODE REVIEW

**Review by weskubo-cgi — CHANGES_REQUESTED** [review]

File: backend/src/routes/applications.js [hunk 1]
```diff
@@ -160,6 +163,20 @@ router.get(
...
 )
+/**
+ * Get an existing Supplementary Application details
+ */
+router.get(
+  '/supplementary/:applicationId/supp',
+  passport.authenticate('jwt', { session: false }),
+  isValidBackendToken,
+  validatePermission(PERMISSIONS.VIEW_APPLICATIONS),
+  [param('applicationId', 'URL param: [applicationId] is required').not().isEmpty()],
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply backend/src/routes/applications.js:206]: done

File: backend/src/routes/applications.js [hunk 2]
```diff
@@ -214,6 +231,21 @@ router.get(
...
 
+/**
+ * Get Approved Supplementary Application PDF by ID
+ */
+router.get(
+  '/supplementary/:applicationId/approved-pdf',
+  passport.authenticate('jwt', { session: false }),
+  isValidBackendToken,
+  validatePermission(PERMISSIONS.VIEW_APPLICATIONS),
+  [param('applicationId', 'URL param: [applicationId] is required').not().isEmpty()],
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply backend/src/routes/applications.js:274]: done

File: backend/src/routes/applications.js [hunk 3]
```diff
@@ -258,3 +290,18 @@ router.delete(
...
+
+/**
+ * Get an Irregular Expense application  using applicationId
+ */
+router.get(
+  '/irregularExpense/:applicationId',
+  passport.authenticate('jwt', { session: false }),
+  isValidBackendToken,
+  validatePermission(PERMISSIONS.VIEW_APPLICATIONS),
+  [param('applicationId', 'URL param: [applicationId] is required').not().isEmpty()],
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply backend/src/routes/applications.js:334]: done

File: backend/src/routes/applications.js [hunk 4]
```diff
@@ -131,6 +136,36 @@ router.get(
...
 
+/**
+ * Get Irregular Expense Application PDF by ID
+ */
+router.get(
+  '/irregular/:applicationId/pdf',
+  passport.authenticate('jwt', { session: false }),
+  isValidBackendToken,
+  validatePermission(PERMISSIONS.VIEW_APPLICATIONS),
+  [param('applicationId', 'URL param: [applicationId] is required').not().isEmpty()],
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply backend/src/routes/applications.js:147]: done

File: backend/src/routes/applications.js [hunk 5]
```diff
@@ -131,6 +136,36 @@ router.get(
...
+
+/**
+ * Get Irregular Expense Application by ID
+ */
+router.get(
+  '/irregular/:applicationId',
+  passport.authenticate('jwt', { session: false }),
+  isValidBackendToken,
+  validatePermission(PERMISSIONS.VIEW_APPLICATIONS),
+  [param('applicationId', 'URL param: [applicationId] is required').not().isEmpty()],
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply backend/src/routes/applications.js:162]: done

File: frontend/src/router/index.js [hunk 6]
```diff
@@ -145,6 +147,24 @@ const router = createRouter({
...
+      path: '/approved-funding/:fundingGuid',
+      name: 'approved-funding',
+      component: ApprovedFundingView,
+      meta: {
+        requiresAuth: true,
+        permission: PERMISSIONS.VIEW_FUNDING_AGREEMENT,
+      },
+    },
+    {
+      path: '/approved-irregular-funding/:fundingGuid',
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply frontend/src/router/index.js:160]: done

File: frontend/src/services/applicationService.js [hunk 7]
```diff
@@ -197,6 +222,40 @@ export default {
...
+      console.log(`Failed to get the approved supp application PDF by application id - ${error}`)
+      throw error
+    }
+  },
+
+  async getIrregularExpensePDF(applicationId) {
+    try {
+      if (!applicationId) return
+      const url = `${ApiRoutes.IRREGULAR_APPLICATIONS}/${applicationId}/pdf`
+      console.log(url)
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply frontend/src/services/applicationService.js:240]: done

File: frontend/src/views/funding/IrregularExpenseFunding.vue [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+<script>
+import alertMixin from '@/mixins/alertMixin'
+import AppButton from '@/components/ui/AppButton.vue'
+import OrganizationHeader from '@/components/organizations/OrganizationHeader.vue'
+import AppBackButton from '@/components/ui/AppBackButton.vue'
+import ApplicationService from '@/services/applicationService'
+import AppPDFViewer from '@/components/ui/AppPDFViewer.vue'
+
+export default {
+  name: 'ApprovedFundingView',
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply frontend/src/views/funding/IrregularExpenseFunding.vue:29]: done

File: backend/src/routes/applications.js [hunk 9]
```diff
@@ -160,6 +163,20 @@ router.get(
     return getSupplementaryApplications(req, res)
   },
 )
+/**
+ * Get an existing Supplementary Application details
+ */
+router.get(
+  '/supplementary/:applicationId/supp',
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply backend/src/routes/applications.js:202]: done

File: backend/src/components/applications.js [hunk 10]
```diff
@@ -164,6 +177,21 @@ async function createSupplementaryApplication(req, res) {
...
+    const applications = []
+    const operation = `ofm_expenses?$filter=(_ofm_application_value eq ${req.params.applicationId} ${buildAppsFilterQuery(req?.query, IrregularExpenseMappings)})`
+    const response = await getOperation(operation)
+
+    response?.value?.forEach((application) => applications.push(new MappableObjectForFront(application, IrregularExpenseMappings).toJSON()))
+
+    return res.status(HttpStatus.OK).json(applications)
+  } catch (e) {
+    log.error(e)
+    return res.status(HttpStatus.INTERNAL_SERVER_ERROR).json(e.data ? e.data : e?.status)
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply backend/src/components/applications.js:191]: done

File: frontend/src/views/applications/ApplicationsHistoryView.vue [hunk 11]
```diff
@@ -156,6 +156,7 @@ export default {
       applications: [],
       supplementaryApplications: [],
       applicationItems: [],
+      irregExpenses: [],
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply frontend/src/views/applications/ApplicationsHistoryView.vue:159]: changed

File: frontend/src/router/index.js [hunk 12]
```diff
@@ -145,6 +146,15 @@ const router = createRouter({
         permission: PERMISSIONS.VIEW_FUNDING_AGREEMENT,
       },
     },
+    {
+      path: '/approved-funding/:fundingGuid',
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply frontend/src/router/index.js:151]: done

File: frontend/src/views/funding/ViewApprovedFunding.vue [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+
+    <h4 class="my-10">Approval Notification For {{ application?.supplementaryTypeDescription }} Allowance</h4>
+
+    <section class="my-5 py-10 grey-div-with-border">
+      <v-card class="mt-5 py-10 pdf-reader">
+        <VuePdfEmbed :source="pdfFile" />
+      </v-card>
+    </section>
+
+    <a style="text-decoration: none" :download="'Approval_Notification_' + application?.supplementaryReferenceNumber" :href="pdfDownloadLink">
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply frontend/src/views/funding/ViewApprovedFunding.vue]: done

File: frontend/src/views/funding/FundingView.vue [hunk 14]
```diff
@@ -29,11 +29,7 @@
       </v-col>
     </v-row>
 
-    <section class="my-5 py-10 grey-div-with-border">
-      <v-card class="mt-5 py-10 pdf-reader">
-        <VuePdfEmbed :source="pdfFile" />
-      </v-card>
-    </section>
+    <AppPDFViewer :pdfFile="pdfFile"></AppPDFViewer>
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply frontend/src/views/funding/FundingView.vue:32]: done

File: frontend/src/views/funding/IrregularExpenseFunding.vue [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+        this.setWarningAlert('PDF Generation is still in progress. Please wait a few minutes before you try again.')
+      } finally {
+        this.loading = false
+      }
+    },
+  },
+}
+</script>
+
+<style>
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply frontend/src/views/funding/IrregularExpenseFunding.vue:65]: done

File: frontend/src/components/ui/AppPDFViewer.vue [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+      required: true,
+      default: () => {
+        return {}
+      },
+    },
+  },
+}
+</script>
+
+<style>
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply frontend/src/components/ui/AppPDFViewer.vue:31]: done

File: backend/src/components/applications.js [hunk 17]
```diff
@@ -144,6 +146,17 @@ async function getSupplementaryApplications(req, res) {
...
 }
 
+async function getSupplementaryApplicationById(req, res) {
+  try {
+    const operation = `ofm_allowances(${req?.params.applicationId})`
+    const response = await getOperation(operation)
+    return res.status(HttpStatus.OK).json(new MappableObjectForFront(response, SupplementaryApplicationMappings).toJSON())
+  } catch (e) {
+    log.error(e)
+    return res.status(HttpStatus.INTERNAL_SERVER_ERROR).json(e.data ? e.data : e?.status)
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply backend/src/components/applications.js:156]: done

File: backend/src/components/applications.js [hunk 18]
```diff
@@ -164,6 +177,21 @@ async function createSupplementaryApplication(req, res) {
   }
 }
 
+async function getIrregularExpenseApplication(req, res) {
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply backend/src/components/applications.js:180]: updated

File: backend/src/components/applications.js [hunk 19]
```diff
@@ -204,6 +232,41 @@ async function getSupplementaryApplicationPDF(req, res) {
   }
 }
 
+async function getIrregularExpensePDF(req, res) {
+  log.info('called')
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply backend/src/components/applications.js:236]: done

File: backend/src/components/applications.js [hunk 20]
```diff
@@ -204,6 +232,41 @@ async function getSupplementaryApplicationPDF(req, res) {
...
+  try {
+    const operation = `ofm_expenses(${req.params.applicationId})/ofm_file`
+    const response = await getOperation(operation)
+    return res.status(HttpStatus.OK).json(response?.value)
+  } catch (e) {
+    log.error(e)
+    return res.status(HttpStatus.INTERNAL_SERVER_ERROR).json(e.data ? e.data : e?.status)
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply backend/src/components/applications.js:243]: done

File: backend/src/components/applications.js [hunk 21]
```diff
@@ -204,6 +232,41 @@ async function getSupplementaryApplicationPDF(req, res) {
...
+  }
+}
+
+async function getIrregularExpenseByID(req, res) {
+  log.info('called')
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply backend/src/components/applications.js:248]: done

File: backend/src/components/applications.js [hunk 22]
```diff
@@ -204,6 +232,41 @@ async function getSupplementaryApplicationPDF(req, res) {
...
+  try {
+    const operation = `ofm_expenses(${req.params.applicationId})`
+    const response = await getOperation(operation)
+    return res.status(HttpStatus.OK).json(new MappableObjectForFront(response, IrregularExpenseMappings).toJSON())
+  } catch (e) {
+    log.error(e)
+    return res.status(HttpStatus.INTERNAL_SERVER_ERROR).json(e.data ? e.data : e?.status)
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply backend/src/components/applications.js:255]: done

File: frontend/src/services/applicationService.js [hunk 23]
```diff
@@ -109,6 +124,16 @@ export default {
     }
   },
 
+  async getSupplementaryApplicationById(supplementaryApplicationId) {
+    try {
+      const response = await ApiService.apiAxios.get(ApiRoutes.SUPPLEMENTARY_APPLICATIONS + '/' + supplementaryApplicationId + '/supp')
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply frontend/src/services/applicationService.js:129]: done

File: frontend/src/services/applicationService.js [hunk 24]
```diff
@@ -54,6 +54,21 @@ export default {
     }
   },
 
+  async getIrregularExpenseApplication(applicationId, filterQuery = undefined) {
+    try {
+      if (!applicationId) return
+      let url = `${ApiRoutes.APPLICATIONS}/irregularExpense/${applicationId}`
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply frontend/src/services/applicationService.js:60]: done

File: frontend/src/services/applicationService.js [hunk 25]
```diff
@@ -54,6 +54,21 @@ export default {
     }
   },
 
+  async getIrregularExpenseApplication(applicationId, filterQuery = undefined) {
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply frontend/src/services/applicationService.js:57]: good point, done

File: backend/src/routes/applications.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (jenbeckett) [reply backend/src/routes/applications.js]: done

File: backend/src/components/applications.js [hunk 26]
```diff
@@ -135,7 +137,7 @@ function buildSupplementaryApplicationsDateQuery(query) {
 async function getSupplementaryApplications(req, res) {
   try {
     const operation = `ofm_allowances?$filter=(${buildSupplementaryApplicationsDateQuery(req?.query)} _ofm_application_value eq ${req?.params.applicationId}
-    ${buildGetSupplementaryApplicationsFilterQuery(req?.query)} )`
+    ${buildAppsFilterQuery(req?.query, SupplementaryApplicationMappings, 'and statuscode ne 2')} )`
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply backend/src/components/applications.js:138]: I think I have a hang on it now - I was having issues getting the 'and' to work nicely - but it should be fixed now

File: backend/src/routes/applications.js [hunk 27]
```diff
@@ -258,3 +290,18 @@ router.delete(
     return deleteEmployeeCertificate(req, res)
   },
 )
+
+/**
+ * Get an Irregular Expense application  using applicationId
+ */
+router.get(
+  '/irregularExpense/:applicationId',
```
[the remark this answers is not in the document]
Comment (jenbeckett) [reply backend/src/routes/applications.js]: done

**Review by weskubo-cgi — APPROVED** [review]

**Pull request merged**